r/R/record-batch.R

```diff
@@ -189,3 +189,59 @@ record_batch <- RecordBatch$create
 
 #' @export
 names.RecordBatch <- function(x) x$names()
+
+#' @export
+rbind.RecordBatch <- function(...) {
+  abort("Use `Table$create()` to combine record batches")
+}
+
+cbind_check_length <- function(inputs, call = caller_env()) {
+  sizes <- map(inputs, function(input) {
+    # TODO: can we somehow implement vec_size for Arrow structures?
+    if (inherits(input, "ArrowTabular") || is.data.frame(input)) {
+      nrow(input)
+    } else {
+      length(input)
+    }
+  })
+  wrong_idx <- which.min(sizes == sizes[[1]] | sizes == 1)
+  if (wrong_idx != 1) {
+    abort(
+      c("Non-scalar inputs must have an equal number of rows.",
+        i = sprintf("..1 has %d, ..%d has %d", sizes[[1]], wrong_idx, sizes[[wrong_idx]])),
+      call = call
+    )
+  }
+}
```

Comment:
I think the meaning here is a little clearer. I used `head()` in the first line because it is robust to length-0 input (whereas `[[1]]` would error) though you also use it in your error message below. Not sure if that's a problem here or not, maybe you can never get in here with `inputs` being empty.

```suggestion
 ok_lengths <- sizes %in% c(head(sizes, 1), 1L)
  if (!all(ok_lengths)) {
    first_bad_one <- which.min(!ok.lengths)
    abort(
      c("Non-scalar inputs must have an equal number of rows.",
        i = sprintf("..1 has %d, ..%d has %d", sizes[[1]], first_bad_one, sizes[[first_bad_one]])),
      call = call
    )
  }
}
```